File: api_server/images/FileSystemImageManager.py
import os
import datetime
from werkzeug.datastructures import FileStorage
from werkzeug.utils import secure_filename
from flask import url_for
from fn.monad import optionable, Option, Empty

from .ImageManager import ImageManager
from config import UPLOAD_DIR
from ImageConverter import ImageConverter


class FileSystemImageManager(ImageManager):
    identifier = 'roselia'
    display_name = 'Roselia'
    description = 'Upload to this site directly.'

    supported_extensions = {'jpg', 'jpeg', 'png', 'bmp', 'webp', 'gif'}

    def __init__(self, upload_dir=None):
        self.upload_dir = upload_dir or UPLOAD_DIR
        assert self.upload_dir, "Please specify UPLOAD_DIR in config.py"

    def is_supported_filename(self, file_name: str):
        return file_name.split('.')[-1].lower() in self.supported_extensions

    def add_image(self, file: FileStorage, file_name: str, convert: str) -> Option:
        if file and self.is_supported_filename(file_name):
            filename = secure_filename(file.filename)
            if not os.path.exists(self.upload_dir):
                os.mkdir(self.upload_dir)
            filename = datetime.datetime.now().strftime("Upload_%Y%m%d%H%M%S_") + filename
            filename = ImageConverter(file, filename, convert).save(self.upload_dir)

            return Option((self.get_url(filename), filename))
        return Empty()

    def list(self):
        file_list = \
            [x for x in os.listdir(self.upload_dir) if
             os.path.isfile(os.path.join(self.upload_dir, x)) and self.is_supported_filename(x)]

        return [{
            'url': url,
            'fileName': name,
            'deleteKey': name,
            'channel': self.identifier
        } for name, url in zip(file_list, map(self.get_url, file_list))]

    def delete_image(self, file_name):
        secure = secure_filename(file_name)
        if secure != file_name:
            return False, 'Do not play tricks.'
        full_path = os.path.join(self.upload_dir, secure)
        success = os.path.exists(full_path)
        if success:
            os.remove(full_path)
            return True, 'Image deleted'
        return False, 'File not found.'

    @classmethod
    def get_url(cls, filename: str) -> str:
        return url_for('uploaded_file', filename=filename)

File: RoseliaCore/Image/FileSystemImageManager.fs
namespace RoseliaCore.Image

open System
open System.Threading.Tasks
open FSharp.Control.Tasks.V2
open RoseliaBlog.RoseliaCore
open RoseliaBlog.RoseliaCore.Image

type FileSystemImageManager() =
    let SupportedExtensions =
        ["jpg"; "jpeg"; "png"; "bmp"; "webp"; "gif"]
        |> Set.ofList
        
    let UploadDir =
        Config.GetConfig<string> "uploadDir"
        |> Option.filter (String.IsNullOrEmpty >> not)
    
    let IsExtensionSupported (ext : string) =
        let extension =
            if ext.StartsWith "." then
                ext.Substring(1)
            else ext
        
        SupportedExtensions.Contains  extension
    let IsFileNameSupported (fileName : string) =
        fileName
        |> IO.Path.GetExtension
        |> IsExtensionSupported
        
    let mutable UrlDelegate : option<string -> string> = None
    
    let GetUrl url =
        UrlDelegate
        |> Option.defaultValue id
        |> fun f -> f url
        
    let SanitizeFileName (name : string) =
        let invalidChars = IO.Path.GetInvalidFileNameChars()
        name.Split(invalidChars, StringSplitOptions.RemoveEmptyEntries)
        |> String.concat "_"
    
    member this.GetStream name =
        name
        |> SanitizeFileName
        |> Option.ofObj
        |> Option.filter IsFileNameSupported
        |> Option.map (fun p -> IO.Path.Join(UploadDir.Value, p))
        |> Option.filter IO.File.Exists
        |> Option.map IO.File.OpenRead
    
    member this.SetUrlBuilder ub =
        UrlDelegate <- (Some ub)
    
    static member TrySetUrlBuilder (manager : IImageManager) (ub : System.Func<string, string>) =
        match manager with
        | :? FileSystemImageManager as fsi ->
            fsi.SetUrlBuilder ub.Invoke
        | _ -> ()
    
    interface IImageManager with
        member this.Identifier = "roselia"
        member this.DisplayName = "Roselia"
        member this.Description = "Upload to this site directly."
        member this.IsDefault = true
        member this.IsEnabled() = UploadDir.IsSome
        member this.AddImageAsync image rawFileName convert =
            task {
                let fileName = SanitizeFileName rawFileName
                if IsFileNameSupported fileName then
                    let prefix = DateTime.Now.ToString "Upload_%Y%m%d%H%M%S_"
                    let newFileName = $"{prefix}{fileName}"
                    use fileStream = IO.File.OpenWrite(IO.Path.Join(UploadDir.Value, newFileName))
                    let! _ = image.CopyToAsync fileStream
                    return Ok {
                        Url = GetUrl newFileName
                        FileName = newFileName
                        DeleteKey = newFileName
                        Channel = (this :> IImageManager).Identifier
                    }
                else return Error "File extension is not supported."
            }
        member this.DeleteImageAsync fileName =
            task {
                let secureFileName = SanitizeFileName fileName
                if secureFileName <> fileName then
                    return Error "Invalid path name"
                else
                    let fullPath = IO.Path.Join(UploadDir.Value, secureFileName)
                    if IO.File.Exists fullPath then
                        IO.File.Delete fullPath
                        return Ok ()
                    else return Error "File not exist."
            }
            
        member this.ListImagesAsync() =
            let files = IO.Directory.GetFiles UploadDir.Value
            files
            |> Seq.map IO.Path.GetFileName 
            |> Seq.filter IsFileNameSupported
            |> Seq.map
                (fun path -> {
                    Url = GetUrl path
                    FileName = path
                    DeleteKey = path
                    Channel = (this :> IImageManager).Identifier
                })
            |> Task.FromResult
